controller/sqlstore/sqlstore.go

```
@@ -0,0 +1,338 @@
+package sqlstore
+
+import (
+	"context"
+	"database/sql"
+	"log"
+	"time"
+
+	_ "github.com/lib/pq" // Import pq driver.
+
+	"github.com/battlesnakeio/engine/controller"
+	"github.com/battlesnakeio/engine/controller/pb"
+	"github.com/battlesnakeio/engine/rules"
+	"github.com/gogo/protobuf/proto"
+	uuid "github.com/satori/go.uuid"
+)
+
+const migrations = `
+CREATE TABLE IF NOT EXISTS locks (
+	key VARCHAR(255) PRIMARY KEY,
+	token VARCHAR(255) NOT NULL,
+	expiry TIMESTAMP NOT NULL
+);
+CREATE TABLE IF NOT EXISTS games (
+	id VARCHAR(255) PRIMARY KEY,
+	value BYTEA
+);
+CREATE TABLE IF NOT EXISTS game_frames (
+	id VARCHAR(255),
+	turn INTEGER,
+	value BYTEA,
+	PRIMARY KEY (id, turn)
+);
+`
+
+// NewSQLStore returns a new store using a postgres database.
+func NewSQLStore(url string) (*Store, error) {
+	db, err := sql.Open("postgres", url)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	if err = db.PingContext(ctx); err != nil {
+		return nil, err
+	}
+
+	_, err = db.ExecContext(ctx, migrations)
+	if err != nil {
+		return nil, err
+	}
+	return &Store{db: db}, nil
+}
+
+// Store represents an SQL store.
+type Store struct {
+	db *sql.DB
+}
+
+// transact is a transaction wrapper, helps avoid failed to close connections.
+func (s *Store) transact(
+	ctx context.Context, txFunc func(*sql.Tx) error) (err error) {
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return
+	}
+	defer func() {
+		if p := recover(); p != nil {
+			if rErr := tx.Rollback(); rErr != nil {
+				log.Printf("rollback failed: %v", rErr)
+			}
+			panic(p) // re-throw panic after Rollback
```

why not return the error here? this could take down the engine completely

---

This is already catching a panic so it's merely a re-throw, I don't want to swallow a panic (at least in this code).

---

ok